Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,602 @@
+* `X_F F` — the subshift defined by forbidding a family of patterns.
+* `SFT F` — a subshift of finite type defined by a finite set of forbidden patterns.
+* `languageOn X U` — the set of patterns of shape `U` obtained by restricting some `x ∈ X`.
+
+## Conventions
+
+We use a **right** action of `G` on configurations:
+`(mulShift g x) h = x (h * g)`.
+In additive notation (e.g. for `ℤ^d`):
+`(addShift v x) u = x (u + v)`.
+
+## Design choice: ambient vs. inner (subshift-relative) viewpoint
+
+All core notions (shift, cylinder, occurrence, language, …) are defined **in the
+ambient full shift** `(G → A)`. A subshift is then a closed, invariant subset,
+bundled as `Subshift A G`. Working inside a subshift is done by restriction.
+
+**Motivation.**
+If cylinders and shifts were defined only *inside* a subshift, local ergonomics
+would improve but global operations would become awkward. For instance, to prove
+that for finite shape `U`:
+
+`languageOn (X ∪ Y) U = languageOn X U ∪ languageOn Y U,`
+
+one must eventually move both sides to the ambient pattern type. Similar issues
+arise for intersections, factors, and products. By contrast, with ambient
+definitions these set-theoretic identities are tautological.
+Thus the file develops the theory ambiently, and subshifts reuse it by restriction.
+
+**Working inside a subshift.**
+For `Y : Subshift A G`, cylinders and occurrence sets *inside `Y`* are simply
+preimages of the ambient ones under the inclusion `Y → (G → A)`. For example:
+
+`{ y : Y | ∀ i ∈ U, (y : G → A) i = (x : G → A) i } = (Subtype.val) ⁻¹’ (cylinder U (x : G → A)).`
+
+Shift invariance guarantees that the ambient shift restricts to `Y`.
+
+**Ergonomics.**
+Thin wrappers (e.g. `Subshift.shift`, `Subshift.cylinder`, `Subshift.languageOn`)
+may be added for convenience. They introduce no new theory and unfold to the
+ambient definitions.
+
+## Namespacing policy
+
+All ambient definitions live under the namespace `SymbolicDynamics.Ambiant`.
+If inner, subshift-relative wrappers are provided, they will be placed in the
+subnamespace `SymbolicDynamics.Inner`. This separation avoids name clashes
+between the two viewpoints, since both may naturally want to reuse names like
+`cylinder`, `shift`, `occursAt`, or `languageOn`.
+
+## Implementation notes
+
+* Openness/closedness results for cylinders and occurrence sets use
+  `[DiscreteTopology A]`. Closedness proofs that enumerate values additionally
+  require `[Fintype A]`, `[DecidableEq A]`, and `[DecidableEq G]` (for `Finset`
+  manipulations and `Function.update`).
+-/
+
+
+noncomputable section
+open Set Topology
+
+namespace SymbolicDynamics
+
+namespace Ambiant
+
+/-! ## Full shift and shift action -/
+
+section ShiftDefinition
+
+variable {A : Type*}
+variable {G : Type*}
+variable [Monoid G]
+
+/-- Right-translation shift:
+* In multiplicative notation: `(mulShift g x) h = x (h * g)`.
+* In additive notation (e.g. for `ℤ^d`): `(addShift v x) u = x (u + v)` -/
```

Comment:
You should have one docstring for the multiplicative notion, only mentioning multiplicative stuff. And another one for the additive notion, only mentioning additive stuff, and given in the `to_additive` call.